File: app/src/main/java/com/example/realxz/realxzdemo/network/adapter/RealXzRxJava2CallAdapter.java
package com.example.realxz.realxzdemo.network.adapter;


import com.example.realxz.realxzdemo.network.ApiResponse;
import com.example.realxz.realxzdemo.network.rx.RealXzBodyFlowable;
import com.example.realxz.realxzdemo.network.rx.RealXzBodyObservable;
import com.example.realxz.realxzdemo.network.rx.RealXzCallExecuteObservable;

import java.lang.reflect.Type;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Observable;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.Response;

/**
 * @author real xz
 * @date 2018/9/7
 */

public class RealXzRxJava2CallAdapter<R> implements CallAdapter<ApiResponse<R>, Object> {
    private final Type responseType;
    private final boolean isFlowable;

    RealXzRxJava2CallAdapter(Type responseType, boolean isFlowable) {
        this.responseType = responseType;
        this.isFlowable = isFlowable;
    }

    @Override
    public Type responseType() {
        return responseType;
    }

    @Override
    public Object adapt(Call<ApiResponse<R>> call) {
        Observable<Response<ApiResponse<R>>> responseObservable = new RealXzCallExecuteObservable<>(call);
        if (isFlowable) {
            return new RealXzBodyFlowable<>(responseObservable.toFlowable(BackpressureStrategy.LATEST));
        } else {
            return new RealXzBodyObservable<>(responseObservable);
        }
    }
}

File: app/src/main/java/com/example/realxz/realxzdemo/network/adapter/RealXzRxJava2CallAdapter.kt
package com.example.realxz.realxzdemo.network.adapter

import com.example.realxz.realxzdemo.network.ApiResponse
import com.example.realxz.realxzdemo.network.rx.RealXzBodyFlowable
import com.example.realxz.realxzdemo.network.rx.RealXzBodyObservable
import com.example.realxz.realxzdemo.network.rx.RealXzCallExecuteObservable
import io.reactivex.BackpressureStrategy
import retrofit2.Call
import retrofit2.CallAdapter
import java.lang.reflect.Type

/**
 * @author real xz
 * @date 2019-06-03
 */
class RealXzRxJava2CallAdapter<R>(private val responseType: Type, private val isFlowable: Boolean) : CallAdapter<ApiResponse<R>, Any> {
    override fun responseType(): Type {
        return responseType
    }

    override fun adapt(call: Call<ApiResponse<R>>): Any {
        val responseObservable = RealXzCallExecuteObservable(call)
        return if (isFlowable) {
            RealXzBodyFlowable(responseObservable.toFlowable(BackpressureStrategy.LATEST))
        } else {
            RealXzBodyObservable(responseObservable)
        }
    }


}